File: bindings/cxx/request-config.cpp
// SPDX-License-Identifier: LGPL-2.1-or-later
// SPDX-FileCopyrightText: 2021 Bartosz Golaszewski <brgl@bgdev.pl>

#include <ostream>
#include <utility>

#include "internal.hpp"

namespace gpiod {

namespace {

request_config_ptr make_request_config()
{
	request_config_ptr config(::gpiod_request_config_new());
	if (!config)
		throw_from_errno("Unable to allocate the request config object");

	return config;
}

} /* namespace */

request_config::impl::impl()
	: config(make_request_config())
{

}

GPIOD_CXX_API request_config::request_config()
	: _m_priv(new impl)
{

}

GPIOD_CXX_API request_config::request_config(request_config&& other) noexcept
	: _m_priv(::std::move(other._m_priv))
{

}

GPIOD_CXX_API request_config::~request_config()
{

}

request_config& request_config::operator=(const request_config& other)
{
	this->_m_priv = other._m_priv;

	return *this;
}

GPIOD_CXX_API request_config& request_config::operator=(request_config&& other) noexcept
{
	this->_m_priv = ::std::move(other._m_priv);

	return *this;
}

GPIOD_CXX_API request_config&
request_config::set_consumer(const ::std::string& consumer) noexcept
{
	::gpiod_request_config_set_consumer(this->_m_priv->config.get(), consumer.c_str());

	return *this;
}

GPIOD_CXX_API ::std::string request_config::consumer() const noexcept
{
	const char* consumer = ::gpiod_request_config_get_consumer(this->_m_priv->config.get());

	return consumer ?: "";
}

GPIOD_CXX_API request_config&
request_config::set_event_buffer_size(::std::size_t event_buffer_size) noexcept
{
	::gpiod_request_config_set_event_buffer_size(this->_m_priv->config.get(),
						     event_buffer_size);

	return *this;
}

GPIOD_CXX_API ::std::size_t request_config::event_buffer_size() const noexcept
{
	return ::gpiod_request_config_get_event_buffer_size(this->_m_priv->config.get());
}

GPIOD_CXX_API ::std::ostream& operator<<(::std::ostream& out, const request_config& config)
{
	::std::string consumer;

	consumer = config.consumer().empty() ? "N/A" : ::std::string("'") + config.consumer() + "'";

	out << "gpiod::request_config(consumer=" << consumer <<
	       ", event_buffer_size=" << config.event_buffer_size() <<
	       ")";

	return out;
}

} /* namespace gpiod */

File: bindings/rust/libgpiod/src/request_config.rs
// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
// SPDX-FileCopyrightText: 2022 Linaro Ltd.
// SPDX-FileCopyrightText: 2022 Viresh Kumar <viresh.kumar@linaro.org>

use std::ffi::{CStr, CString};
use std::os::raw::c_char;
use std::str;

use super::{Error, OperationType, Result, gpiod};

/// Request configuration objects
///
/// Request config objects are used to pass a set of options to the kernel at
/// the time of the line request. The mutators don't return error values. If the
/// values are invalid, in general they are silently adjusted to acceptable
/// ranges.

#[derive(Debug, Eq, PartialEq)]
pub struct Config {
    pub(crate) config: *mut gpiod::gpiod_request_config,
}

// SAFETY: Config models a wrapper around an owned gpiod_request_config and may
// be safely sent to other threads.
unsafe impl Send for Config {}

impl Config {
    /// Create a new request config object.
    pub fn new() -> Result<Self> {
        // SAFETY: The `gpiod_request_config` returned by libgpiod is guaranteed to live as long
        // as the `struct Config`.
        let config = unsafe { gpiod::gpiod_request_config_new() };
        if config.is_null() {
            return Err(Error::OperationFailed(
                OperationType::RequestConfigNew,
                errno::errno(),
            ));
        }

        Ok(Self { config })
    }

    /// Set the consumer name for the request.
    ///
    /// If the consumer string is too long, it will be truncated to the max
    /// accepted length.
    pub fn set_consumer(&mut self, consumer: &str) -> Result<&mut Self> {
        let consumer = CString::new(consumer).map_err(|_| Error::InvalidString)?;

        // SAFETY: `gpiod_request_config` is guaranteed to be valid here.
        unsafe {
            gpiod::gpiod_request_config_set_consumer(
                self.config,
                consumer.as_ptr() as *const c_char,
            )
        }

        Ok(self)
    }

    /// Get the consumer name configured in the request config.
    pub fn consumer(&self) -> Result<&str> {
        // SAFETY: The string returned by libgpiod is guaranteed to live as long
        // as the `struct Config`.
        let consumer = unsafe { gpiod::gpiod_request_config_get_consumer(self.config) };
        if consumer.is_null() {
            return Err(Error::OperationFailed(
                OperationType::RequestConfigGetConsumer,
                errno::errno(),
            ));
        }

        // SAFETY: The string is guaranteed to be valid here by the C API.
        unsafe { CStr::from_ptr(consumer) }
            .to_str()
            .map_err(Error::StringNotUtf8)
    }

    /// Set the size of the kernel event buffer for the request.
    pub fn set_event_buffer_size(&mut self, size: usize) -> &mut Self {
        // SAFETY: `gpiod_request_config` is guaranteed to be valid here.
        unsafe { gpiod::gpiod_request_config_set_event_buffer_size(self.config, size) }

        self
    }

    /// Get the edge event buffer size setting for the request config.
    pub fn event_buffer_size(&self) -> usize {
        // SAFETY: `gpiod_request_config` is guaranteed to be valid here.
        unsafe { gpiod::gpiod_request_config_get_event_buffer_size(self.config) }
    }
}

impl Drop for Config {
    /// Free the request config object and release all associated resources.
    fn drop(&mut self) {
        // SAFETY: `gpiod_request_config` is guaranteed to be valid here.
        unsafe { gpiod::gpiod_request_config_free(self.config) }
    }
}
